src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopies = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      stateCopy = addProperties(stateCopy, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      stateCopy = removeProperties(stateCopy, action.keysToRemove);
+    } else if (action.type === 'clear') {
+      stateCopy = clear(stateCopy);
+    }
+
+    stateCopies.push({ ...stateCopy });
+  }
+
+  return stateCopies;
+}
+
+function addProperties(stateCopy, extraData) {
+  return {
+    ...stateCopy,
+    ...extraData,
+  };
+}
+
+function removeProperties(stateCopy, keysToRemove) {
+  const updatedCopy = { ...stateCopy };
+
+  for (const key of keysToRemove) {
+    delete updatedCopy[key];
+  }
+
+  return updatedCopy;
+}
```

Comment:
Similar to the 'addProperties' function, the name 'removeProperties' could be improved by renaming it to 'createStateWithRemovedProperties' to indicate that it creates a new state rather than mutating the existing one.